sky/provision/fluidstack/instance.py

```diff
@@ -0,0 +1,271 @@
+"""FluidStack instance provisioning."""
+import time
+from typing import Any, Dict, List, Optional
+
+from sky import authentication as auth
+from sky import sky_logging
+from sky import status_lib
+from sky.provision import common
+from sky.provision.fluidstack import fluidstack_utils as utils
+from sky.utils import command_runner
+from sky.utils import common_utils
+from sky.utils import subprocess_utils
+from sky.utils import ux_utils
+
+_GET_INTERNAL_IP_CMD = ('ip -4 -br addr show | grep UP | grep -Eo '
+                        r'"(10\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)|'
+                        r'172\.(1[6-9]|2[0-9]|3[0-1]))\.(25[0-5]|'
+                        r'2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|'
+                        r'2[0-4][0-9]|[01]?[0-9][0-9]?)"')
+POLL_INTERVAL = 5
+
+logger = sky_logging.init_logger(__name__)
+
+
+def get_internal_ip(node_info: Dict[str, Any]) -> None:
+    node_info['internal_ip'] = node_info['ip_address']
+    runner = command_runner.SSHCommandRunner(
+        node_info['ip_address'],
+        ssh_user=node_info['capabilities']['default_user_name'],
+        ssh_private_key=auth.PRIVATE_SSH_KEY_PATH)
+    result = runner.run(_GET_INTERNAL_IP_CMD,
+                        require_outputs=True,
+                        stream_logs=False)
+    # rc, stdout, stderr = result
+    if result[0] != 0:
+        logger.error('Failed get obtain private IP from node')
+    # subprocess_utils.handle_returncode(rc,
+    #                                  _GET_INTERNAL_IP_CMD,
+    #                                  'Failed get obtain private IP from node',
+    #                                   stderr=stdout + stderr)
```

Comment:
We should probably raise here to fail early when we fail to get internal IP. Do we want to uncomment this?

Reply:
In some DCs, due to the virtualisation/configuration, `ip -4 -br addr show | grep UP` only shows the external IP address, which fails the regex.

Reply:
Ahh, I see. Could we add a comment here?
```
# Some DCs do not have internal IPs and can fail when getting the IP. We set the `internal_ip` to the same as
# external IP. It should be fine as the `ray cluster` will also get and use that external IP in that case.
```

Reply:
Just tried the region without the internal IP, and it seems working correctly. Thanks for fix here.

nit: it would be good to change `logger.error` above to `logger.debug` as well, since whether or not the internal IP presents here will not affect the end-user's usage of the VM. : )